File: go-backend-interview-prep/10_design_patterns_in_go/10_pipeline_and_handlers_chain/main.go
package main

import (
	"errors"
	"fmt"
	"strings"
	"time"
)

// --- Message ---

type Message struct {
	To        string
	Body      string
	Timestamp time.Time
	Metadata  map[string]string
}

// --- Handler interface ---

type Handler interface {
	Handle(msg *Message) error
}

// --- Pipeline chains handlers ---

type Pipeline struct {
	handlers []Handler
}

func NewPipeline(handlers ...Handler) *Pipeline {
	return &Pipeline{handlers: handlers}
}

func (p *Pipeline) Run(msg *Message) error {
	for _, h := range p.handlers {
		if err := h.Handle(msg); err != nil {
			return err // short-circuit on error
		}
	}
	return nil
}

// --- Concrete handlers ---

// ValidateHandler checks required fields.
type ValidateHandler struct{}

func (v *ValidateHandler) Handle(msg *Message) error {
	fmt.Println("  [VALIDATE] checking message...")
	if msg.To == "" {
		return errors.New("validation: 'to' is required")
	}
	if msg.Body == "" {
		return errors.New("validation: 'body' is required")
	}
	if !strings.Contains(msg.To, "@") {
		return errors.New("validation: 'to' must be an email")
	}
	fmt.Println("  [VALIDATE] OK")
	return nil
}

// EnrichHandler adds metadata.
type EnrichHandler struct{}

func (e *EnrichHandler) Handle(msg *Message) error {
	msg.Timestamp = time.Now()
	if msg.Metadata == nil {
		msg.Metadata = make(map[string]string)
	}
	msg.Metadata["enriched"] = "true"
	msg.Metadata["priority"] = "normal"
	fmt.Printf("  [ENRICH]   Added timestamp + metadata\n")
	return nil
}

// LogHandler logs the message.
type LogHandler struct{}

func (l *LogHandler) Handle(msg *Message) error {
	fmt.Printf("  [LOG]      to=%s body=%q metadata=%v\n",
		msg.To, msg.Body, msg.Metadata)
	return nil
}

// SendHandler delivers the message.
type SendHandler struct{}

func (s *SendHandler) Handle(msg *Message) error {
	fmt.Printf("  [SEND]     Delivered to %s at %s\n",
		msg.To, msg.Timestamp.Format("15:04:05"))
	return nil
}

func main() {
	fmt.Println("=== Pipeline and Handlers Chain ===")
	fmt.Println()

	pipeline := NewPipeline(
		&ValidateHandler{},
		&EnrichHandler{},
		&LogHandler{},
		&SendHandler{},
	)

	// Valid message
	fmt.Println("--- Valid message ---")
	msg1 := &Message{To: "alice@example.com", Body: "Your order shipped!"}
	if err := pipeline.Run(msg1); err != nil {
		fmt.Printf("  PIPELINE ERROR: %v\n", err)
	}
	fmt.Println()

	// Invalid: missing email
	fmt.Println("--- Invalid message (no @) ---")
	msg2 := &Message{To: "bob", Body: "test"}
	if err := pipeline.Run(msg2); err != nil {
		fmt.Printf("  PIPELINE ERROR: %v\n", err)
	}
	fmt.Println()

	// Invalid: empty body
	fmt.Println("--- Invalid message (empty body) ---")
	msg3 := &Message{To: "charlie@example.com", Body: ""}
	if err := pipeline.Run(msg3); err != nil {
		fmt.Printf("  PIPELINE ERROR: %v\n", err)
	}
	fmt.Println()

	fmt.Println("Key: validate -> enrich -> log -> send. Error = short-circuit.")
	fmt.Println("     Each handler does one thing. Add/remove steps freely.")
}

File: go-backend-interview-prep/10_design_patterns_in_go/10_pipeline_and_handlers_chain/main.py
"""Pipeline and handlers chain -- Python equivalent of the Go example.

Message flows through ordered handlers: validate -> enrich -> log -> send.
"""

from datetime import datetime


class Message:
    def __init__(self, to: str, body: str):
        self.to = to
        self.body = body
        self.timestamp: datetime | None = None
        self.metadata: dict[str, str] = {}


class Pipeline:
    def __init__(self, *handlers):
        self.handlers = handlers

    def run(self, msg: Message) -> str | None:
        """Returns error string on failure, None on success."""
        for h in self.handlers:
            err = h.handle(msg)
            if err:
                return err  # short-circuit
        return None


# --- Concrete handlers ---

class ValidateHandler:
    def handle(self, msg: Message) -> str | None:
        print("  [VALIDATE] checking message...")
        if not msg.to:
            return "validation: 'to' is required"
        if not msg.body:
            return "validation: 'body' is required"
        if "@" not in msg.to:
            return "validation: 'to' must be an email"
        print("  [VALIDATE] OK")
        return None


class EnrichHandler:
    def handle(self, msg: Message) -> str | None:
        msg.timestamp = datetime.now()
        msg.metadata["enriched"] = "true"
        msg.metadata["priority"] = "normal"
        print("  [ENRICH]   Added timestamp + metadata")
        return None


class LogHandler:
    def handle(self, msg: Message) -> str | None:
        print(f"  [LOG]      to={msg.to} body={msg.body!r} metadata={msg.metadata}")
        return None


class SendHandler:
    def handle(self, msg: Message) -> str | None:
        ts = msg.timestamp.strftime("%H:%M:%S") if msg.timestamp else "?"
        print(f"  [SEND]     Delivered to {msg.to} at {ts}")
        return None


def main() -> None:
    print("=== Pipeline and Handlers Chain ===")
    print()

    pipeline = Pipeline(
        ValidateHandler(),
        EnrichHandler(),
        LogHandler(),
        SendHandler(),
    )

    # Valid message
    print("--- Valid message ---")
    err = pipeline.run(Message("alice@example.com", "Your order shipped!"))
    if err:
        print(f"  PIPELINE ERROR: {err}")
    print()

    # Invalid: no @
    print("--- Invalid message (no @) ---")
    err = pipeline.run(Message("bob", "test"))
    if err:
        print(f"  PIPELINE ERROR: {err}")
    print()

    # Invalid: empty body
    print("--- Invalid message (empty body) ---")
    err = pipeline.run(Message("charlie@example.com", ""))
    if err:
        print(f"  PIPELINE ERROR: {err}")
    print()

    print("Key: validate -> enrich -> log -> send. Error = short-circuit.")
    print("     Each handler does one thing. Add/remove steps freely.")


if __name__ == "__main__":
    main()
